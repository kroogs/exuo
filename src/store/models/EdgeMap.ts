/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import { IAnyModelType, IAnyType, Instance, types } from 'mobx-state-tree'

type EdgeTypeFn = () => IAnyType

export const edgeMapFactory = (getEdgeType: EdgeTypeFn): IAnyModelType =>
  types
    .model('EdgeMap', {
      edgeMap: types.map(types.array(types.reference(types.late(getEdgeType)))),
    })
    .actions(self => ({
      hasEdge(tag: string, target: Instance<IAnyModelType>): boolean {
        return Boolean(self.edgeMap.get(tag)?.includes(target))
      },
    }))
    .actions(self => ({
      addEdge(tag: string, target: Instance<IAnyModelType>): void {
        if (!self.edgeMap.has(tag)) {
          self.edgeMap.set(tag, [])
        }

        self.edgeMap.get(tag)?.push(target)
      },

      getEdgeTag(tag: string): void | Array<Instance<IAnyModelType>> {
        return self.edgeMap.get(tag)
      },

      removeEdge(tag: string, target: Instance<IAnyModelType>): void {
        if (!self.hasEdge(tag, target)) {
          throw Error(`Node '${self}' has no '${tag}' edge for '${target}'`)
        }

        self.edgeMap.get(tag)?.remove(target)
      },
    }))

// export const getEdgeMapModels = (): void => {
//   return
// }
