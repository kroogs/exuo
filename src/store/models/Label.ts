/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import { types as t } from 'mobx-state-tree'

export const Label = t
  .model('Label', {
    label: t.maybe(t.string),
  })
  .actions(self => ({
    setLabel(label: string) {
      self.label = label
    },
  }))

export default Label
