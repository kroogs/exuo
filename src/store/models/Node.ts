/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import { IAnyModelType, IAnyType, types as t } from 'mobx-state-tree'

import { edgeMapFactory } from './EdgeMap'

type EdgeTypeFn = () => IAnyType

export const nodeFactory = (
  getEdgeType: EdgeTypeFn,
  composeModels: Array<IAnyModelType | EdgeTypeFn> = [],
): IAnyModelType =>
  t
    .compose(
      // @ts-ignore
      ...[
        edgeMapFactory,
        ...composeModels,
        t.model({
          id: t.identifier,
        }),
      ].map(item => (typeof item === 'function' ? item(getEdgeType) : item)),
    )
    .named('Node') // TODO composed names? ("Node+Label+Banana"?)

export const Node = nodeFactory(() => Node)
