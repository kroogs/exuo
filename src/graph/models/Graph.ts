/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { Instance, SnapshotIn, IAnyModelType } from 'mobx-state-tree'

import { persist } from 'store/persist'

import { graphFactory } from './factories'
import { Node, Config, Note } from './Node'

export const Graph = graphFactory({ Node, Config, Note })
  // Modes
  .actions(self => ({
    setActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')
      if (activeModes && !activeModes?.includes(mode)) {
        activeModes.push(mode)
      }
    },

    unsetActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')
      if (activeModes?.includes(mode)) {
        activeModes.remove(mode)
      }
    },

    toggleActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')

      if (!activeModes) {
        return
      }

      if (activeModes.includes(mode)) {
        activeModes.remove(mode)

        // TODO mode logic doesn't belong here
        if (mode === 'select') {
          self.clearSelectedNodes()
        }
      } else {
        activeModes.push(mode)
      }
    },
  }))

  // Selection
  .actions(self => ({
    clearSelectedNodes() {
      self.Config.get('system').get('selectedNodes')?.clear()
    },

    deleteSelectedNodes() {
      const selectedNodes = self.Config.get('system')?.get('selectedNodes')
      selectedNodes.forEach((id: string) => self.Node.delete(id))
      self.clearSelectedNodes()
      if (self.selectedNodes.length === 0) {
        self.unsetActiveMode('select')
      }
    },

    toggleSelectNode(node: Instance<typeof Node>) {
      const selectedNodes = self.Config.get('system').get('selectedNodes')
      if (selectedNodes?.includes(node.id)) {
        selectedNodes?.remove(node.id)
      } else {
        selectedNodes?.push(node.id)
      }

      if (selectedNodes.length === 0) {
        self.toggleActiveMode('select')
      }
    },

    // selectNode(node: Instance<typeof Node>) {

    // },

    // deselectNode(node: Instance<typeof Node>) {

    // },

    selectAllChildNodes(node: Instance<typeof Node>) {
      node.edgeMap
        .get('child')
        .forEach((child: Instance<typeof Node>) => console.log('child'))
    },

    // deselectAllChildNodes(node: Instance<typeof Node>) {

    // },
  }))

  .actions(self => ({
    afterCreate() {
      persist(self).then(() => {
        if (!self.Config.has('system')) {
          const root = self.createNode('Node', { label: 'Exuo' })
          self.createNode('Config', {
            id: 'system',
            items: {
              rootNodeId: root.id,
              activeModes: [],
              selectedNodes: [],
            },
          })
          self.createNode('Config', {
            id: 'user',
            items: {
              global: {
                dividers: false,
              },
              lists: {
                checkbox: false,
                dividers: false,
                showChildCount: true,
                showEdgeChips: false,
              },
            },
          })
        }
      })
    },

    createChild<T extends IAnyModelType>(
      node: Instance<typeof Node>,
      edgeProps: SnapshotIn<T>,
      edgeType = 'Node',
    ): Instance<typeof Node> {
      const child = self.createNode(edgeType, edgeProps)

      node.addEdge('child', child)
      child.addEdge('parent', node)

      return child
    },

    deleteNode(node: Instance<typeof Node>) {
      self.Node.delete(node.id)
    },

    toggleGlobalDividerSetting() {
      const value = self.Config.get('user').get('global')?.get('dividers')
      self.Config.get('user').get('global')?.set('dividers', !value)
    },

    toggleListCheckboxSetting() {
      const value = self.Config.get('user').get('lists')?.get('checkbox')
      self.Config.get('user').get('lists')?.set('checkbox', !value)
    },
  }))
  .views(self => ({
    get rootNode() {
      const config = self.Config.get('system')
      return config && self.Node.get(config.get('rootNodeId'))
    },

    get activeModes() {
      const config = self.Config.get('system')
      return config?.get('activeModes')
    },

    get selectedNodes() {
      return self.Config.get('system')?.get('selectedNodes')
    },

    get listCheckboxSetting() {
      return self.Config.get('user')?.get('lists')?.get('checkbox')
    },
  }))
