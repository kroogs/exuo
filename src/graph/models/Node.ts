/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types, IAnyType } from 'mobx-state-tree'

import { nodeFactory, edgeMapFactory } from 'graph/factories'
import * as models from 'graph/models'

export const Config = nodeFactory(models.Config)
export const Node = nodeFactory([
  models.Log,
  edgeMapFactory(() =>
    types.union(
      types.late((): IAnyType => Node),
      Config,
    ),
  ),
])
  .props({
    label: types.maybe(types.string),
  })
  .actions(self => ({
    afterCreate() {
      self.logEvent('create')
    },
  }))
  .views(self => ({
    get childCount(): number {
      return self.edgeMap.get('child')?.length ?? 0
    },

    get createDate(): Date {
      return self.events[0]?.date
    },
  }))
