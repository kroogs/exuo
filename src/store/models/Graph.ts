/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import {
  Instance,
  onPatch,
  getSnapshot,
  getMembers,
  applySnapshot,
} from 'mobx-state-tree'
import Dexie from 'dexie'

import { graphFactory } from 'store/graph'
import { Node, Config } from './Node'

async function persist(graph: Instance<typeof Graph>): Promise<void> {
  const db = new Dexie('default')
  const tableNames = Object.keys(getMembers(graph).properties)
  const tableConfig = Object.fromEntries(tableNames.map(name => [name, 'id']))

  db.version(1).stores(tableConfig)

  return Promise.all(
    tableNames.map(name =>
      db
        .table(name)
        .toArray()
        .then(rows => [name, Object.fromEntries(rows.map(r => [r.id, r]))]),
    ),
  ).then(resolved => {
    applySnapshot(graph, Object.fromEntries(resolved))
    onPatch(graph, patch => {
      const [typeName, id] = patch.path.split('/').slice(1)
      if (patch.op === 'add' || patch.op === 'replace') {
        db.table(typeName).put(getSnapshot(graph[typeName].get(id)))
      }
    })
  })
}

async function initialize(graph: Instance<typeof Graph>): Promise<void> {
  if (graph.Config.has('graph')) {
    return
  }

  const root = graph.createNode('Node', { label: 'Root' })
  graph.createNode('Config', {
    id: 'graph',
    items: { rootNodeId: root.id },
  })
}

export const Graph = graphFactory({ Node, Config }).actions(self => ({
  afterCreate() {
    const adapters = [persist, initialize]
    adapters
      .reduce(async (prev, next) => {
        await prev
        return next(self)
      }, Promise.resolve())
      .catch(error => {
        throw Error(`Graph adapter error: ${error}`)
      })
  },
}))
