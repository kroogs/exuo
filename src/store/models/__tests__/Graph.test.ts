/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import {
  types as t,
  getType,
  getSnapshot,
  IAnyModelType,
  isType,
} from 'mobx-state-tree'

import { Graph, graphFactory } from '../Graph'
import { nodeFactory } from '../Node'
import { edgeMapFactory } from '../EdgeMap'
import Label from '../Label'

describe('#graphFactory', () => {
  test('supports a custom base Node model', () => {
    const Node = nodeFactory(() => Node, [t.model({ custom: t.string })])
    const GraphModel = graphFactory({ Node })
    expect(
      getSnapshot(
        GraphModel.create({
          nodesById: { one: { id: 'one', custom: 'one' } },
        }),
      ),
    ).toStrictEqual({
      nodesById: { one: { id: 'one', custom: 'one', edgeMap: {} } },
      typesById: {},
    })
  })

  test('supports relationships with custom models', () => {
    const Node = nodeFactory(() => t.union(Node, LabelNode))
    const LabelNode = nodeFactory(() => t.union(Node, LabelNode), [Label])
    const GraphModel = graphFactory({ Node, Label: LabelNode })

    const graph = GraphModel.create()

    const one = graph.createNode()
    const two = graph.createNode()

    two.addEdge('side', one)
    expect(graph.nodesById.get(two.id).edgeMap.get('side')?.[0]).toBe(one)

    const potato = graph.createNode('Label', { label: 'potato' })
    one.addEdge('item', potato)
    expect(graph.nodesById.get(one.id).edgeMap.get('item')?.[0]).toBe(potato)

    potato.addEdge('inside', potato)
    expect(potato.edgeMap.get('inside')?.[0]).toBe(potato)
  })

  test.skip('#createTypeConfig creates and stores TypeConfig instances', () => {
    // graph.createTypeConfig([
    //   {
    //     name: 'Test',
    //     props: [['label', 'string']],
    //   },
    // ])
    return
  })

  test.skip('#recompose produces a new graph with configured types', () => {
    const graph = Graph.create({
      nodesById: {
        one: {
          id: 'one',
        },
      },
      typesById: {
        Custom: {
          name: 'Custom',
          compose: ['Node'],
          props: [['custom', 'string']],
        },
      },
    })

    const newGraph = graph.recompose()
    const custom = newGraph.createNode('Custom', { custom: 'sure' })
    newGraph.nodesById.get('one').addEdge('out', custom)
  })
})
