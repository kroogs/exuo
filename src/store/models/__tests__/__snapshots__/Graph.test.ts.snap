// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph #edgeMapFactory #removeEdge throws an exception when an edge ref isn't found 1`] = `"Node 'AnonymousModel@/items/one(id: one)' has no 'next' edge for 'AnonymousModel@/items/one(id: one)'"`;
